api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -712,6 +712,10 @@ public T map(Types.MapType map, Supplier<T> keyResult, Supplier<T> valueResult)
       return null;
     }
 
+    public T variant(Types.VariantType variant) {
+      throw new UnsupportedOperationException("Unsupported type: variant");
```

Comment:
Looks like this is not yet implemented by `CheckCompatibility`, `FixupTypes`, or `FlinkFixupTypes`.

It is also not implemented by `PruneColumnsWithReordering` but that doesn't appear to be referenced anywhere. I opened #12258 to remove these.

Reply:
Added the implementations other than PruneColumnsWithoutReordering. For FixupTypes for Spark/Flink, there is no fixup for Variant and I didn't add coverage for it, unless we just want to test out the exception is not thrown.